BetterShardsBukkit/src/main/java/vg/civcraft/mc/bettershards/database/DatabaseManager.java

```diff
@@ -784,23 +795,12 @@ private Portal getPortalData(String name, LocationWrapper first, LocationWrapper
 			int specialId = set.getInt("portal_type"); // determine the type of portal.
 			String serverName = set.getString("server_name");
 			String partner = set.getString("partner_id");
-			boolean currentServer = serverName.equals(MercuryAPI.serverName());
-			switch (specialId) {
-			case 0:
-				CuboidPortal p = new CuboidPortal(name, first.getFakeLocation(), second.getFakeLocation(), partner, currentServer);
-				p.setServerName(serverName);
-				return p;
-			case 1:
-				WorldBorderPortal wb = new WorldBorderPortal(name, partner, currentServer, first, second);
-				wb.setServerName(serverName);
-				return wb;
-			case 2:
-				CircularPortal cp = new CircularPortal(name, partner, currentServer, first.getFakeLocation(), second.getFakeLocation());
-				cp.setServerName(serverName);
-				return cp;
-			default:
+			PortalLoadEvent event = new PortalLoadEvent(name, specialId, serverName, partner,
+					first, second);
+			Bukkit.getPluginManager().callEvent(event);
+			if (event.getPortal() == null) // No Portal was set.
 				return null;
-			}
+			return event.getPortal();
```

Comment:
Are you sure you want the API to offer changing the portal being loaded? It means that if I request a portal with name "XYZ" from the database, I might get anything else, because of listeners interferring. You should be careful to separate database storage and gameplay mechanics here.